### PR TITLE
Rework CONTRIBUTING guidelines

### PR DESCRIPTION
**Pull request by kchibisov**
Give better recommendations for contributors on how to propose start the work, write commit messages, and handle review by explicitly stating expectations, however make it clear that maintainers will help you if you have issues.

Write down the guidelines for maintainers on how they should handle contributions and interact with other maintainers to help maintaining good code quality level by listing maintainers responsibilities and providing suggestions on how to handle various situations.

### CODE REVIEW

File: CONTRIBUTING.md
Comment (notgull) [inline CONTRIBUTING.md:26]: I think that a link here to the Git standards this is taken from would be appreciated.

File: CONTRIBUTING.md
Comment (notgull) [inline CONTRIBUTING.md:146]: This paragraph seems unnecessary and overly restrictive, so some context would be appreciated.

**Review by notgull — CHANGES_REQUESTED** [review]
It should also be mentioned here that Markdown files should be wrapped to 80 characters.

File: CONTRIBUTING.md
Comment (kchibisov) [reply CONTRIBUTING.md:146]: I mean, it's a regular process of evolving. It's present in mostly every project I worked in.

File: CONTRIBUTING.md
Comment (kchibisov) [reply CONTRIBUTING.md:26]: 50 characters is not that hard to google, but the source is a bit vague on that. I'd add that multiple editors (nvim, helix) do that sort of indication by default. The 72 chars body is from kernel and in general a lot of inspired by kernel devs, since they know how to write commits. 

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#describe-your-changes

**Comment by kchibisov** [comment]
>It should also be mentioned here that Markdown files should be wrapped to 80 characters.

I was thinking on adding `.editorconfig` for that.

**Comment by notgull** [comment]
> I was thinking on adding `.editorconfig` for that.

While this would be a good solution I would prefer it if it was stated explicitly here.

File: CONTRIBUTING.md
Comment (notgull) [reply CONTRIBUTING.md:146]: I guess what I'm asking is, what bad behavior is this intended to stop? I feel like this is a solution in search of a problem.

File: CONTRIBUTING.md
Comment (notgull) [reply CONTRIBUTING.md:26]: Having this linked in this document would be nice

**Review by daxpedda — CHANGES_REQUESTED** [review]
I think in general this looks good, I would probably like to proof read it but otherwise I believe this is a great improvement.

---

Personally I'm not in favor of documenting changes in commit messages. Documenting changes in PRs has to be done anyway, as it is required to actually get anything merged, otherwise nobody would know whats going on. So to me this is duplicating that work.

I do believe that documenting changes in commit messages is an improvement, but it comes with significant drawbacks. Most notably its a lot more work and according to the new guidelines, that work has to be taken on by maintainers for contributors. Those tradeoffs ought to be weighed and discussed.

Additionally, I would like to point out that this isn't a particularly popular way to do things anymore, e.g. apart from Linux and security related projects I've not seen it often in the Rust open source ecosystem. I believe that today, in general, documenting changes has moved from commit messages to PR's and issues, which I personally don't mind.

I also believe that squashing is fine as long as its not worth to preserve the commit history, which it often isn't in Winit, especially for smaller changes (apart from the fact that commit history is still preserved in the PR itself).

Good commit messages are good though and have minimal tradeoff, I don't see a reason to not have good commit messages. But I do think that even 60 characters is a very traditional limit that doesn't really make sense with modern screens and UIs anymore. Again, we should discuss the tradeoffs here.

I want to clarify that I do actually believe that clearly documenting changes in the commit message is always going to be an improvement. But every improvement has tradeoffs that we should weigh and discuss.

Imo, the Rust open source ecosystem in general doesn't require documenting changes in commit messages (not even the Rust project itself), probably because its a much newer ecosystem, therefor I don't see why we should be different here considering the tradeoffs.

**Comment by kchibisov** [comment]
>I do believe that documenting changes in commit messages is an improvement, but it comes with significant drawbacks. Most notably its a lot more work and according to the new guidelines, that work has to be taken on by maintainers for contributors. Those tradeoffs ought to be weighed and discussed.

I already do that myself for each PR I merge on the span of 3 years. I don't mind continue doing so, it usually takes like a few minutes. Keep in mind that it's recommended to write correct messages, but to not demotivate contributors, I don't mind fixing them myself.

>I also believe that squashing is fine as long as its not worth to preserve the commit history, which it often isn't in Winit, especially for smaller changes (apart from the fact that commit history is still preserved in the PR itself).

Squashing is fine, the thing is that this option can't be used reliably by some parties, once there will be indication that we can write commit messages so things look consistent, I'd bring it back.

>Good commit messages are good though and have minimal tradeoff, I don't see a reason to not have good commit messages. But I do think that even 60 characters is a very traditional limit that doesn't really make sense with modern screens and UIs anymore. Again, we should discuss the tradeoffs here.

Well, it's the topic line, brevity is more important. If you need more than you likely need to split commit into multiples.

>Imo, the Rust open source ecosystem in general doesn't require documenting changes in commit messages (not even the Rust project itself), probably because its a much newer ecosystem, therefor I don't see why we should be different here considering the tradeoffs.

That sounds like their issue and the fallout of using bors. Though, people generally document why they make a particular change. The thing is that you don't really document, but rather express motivation for change, the documentation should belong in code.

File: CONTRIBUTING.md
Comment (madsmtm) [inline CONTRIBUTING.md:115]: I acknowledge that this may lead to decreased code quality, but I believe that requiring the problem to be solved in general is the wrong approach for Winit. While the project is a few years old, it's still young enough that we don't yet know how to do everything! And the absolute best way to get knowledge is to let people experiment with the code!

As an example, https://github.com/rust-windowing/winit/pull/3499 is solving a short-term goal, but I believe it'll also help guide how the API will have to look like in the future.

Another example is the file open APIs on macOS, which got rejected partly because of technical difficulties in making it work right, but also because we wanted to find some solution to https://github.com/rust-windowing/winit/issues/2120, which has been open for over two years with only recently a solution on the horizon.

My opinion is that we need to get all of the sub-par code into Winit, so that we can iterate on it in there instead of in a PR that will have to be rebased over an over (keyboard v2).

**Review by madsmtm — CHANGES_REQUESTED** [review]
Heh, I've been stressing out the whole week because I wanted to write down some of the things we talked about last meeting so that we could continue the discussion, and bam, I should've just looked on GH, you already did it ;)

So thank you, it will certainly help us in aligning our goals!

File: CONTRIBUTING.md
Comment (kchibisov) [reply CONTRIBUTING.md:115]: >I acknowledge that this may lead to decreased code quality, but I believe that requiring the problem to be solved in general is the wrong approach for Winit. While the project is a few years old, it's still young enough that we don't yet know how to do everything! And the absolute best way to get knowledge is to let people experiment with the code!

it's like 10 years old at this point and it was inside glutin at some point.

>As an example, https://github.com/rust-windowing/winit/pull/3499 is solving a short-term goal, but I believe it'll also help guide how the API will have to look like in the future.

This is not how the API will look though. The trait extension is _fine_, but it was not proposed there.

>Another example is the file open APIs on macOS, which got rejected partly because of technical difficulties in making it work right, but also because we wanted to find some solution to https://github.com/rust-windowing/winit/issues/2120, which has been open for over two years with only recently a solution on the horizon.

Yes, the solution is merged and can be used to implement the said API already.

>My opinion is that we need to get all of the sub-par code into Winit, so that we can iterate on it in there instead of in a PR that will have to be rebased over an over (keyboard v2).

No, the subpar code will stay with us forever, that's how it works unfortunately. Once you land it it'll be pretty much like that for a long time. There're exceptions, but it's how it is, unfortunately. You should never justify not maintainable code.

File: CONTRIBUTING.md
Comment (kchibisov) [reply CONTRIBUTING.md:159]: I'm not sure it does? It requires you to be collaborator.

File: CONTRIBUTING.md
Comment (kchibisov) [reply CONTRIBUTING.md:150]: If you have a better name sure, but it's a subsection in maintainers section extending on the contributors section.

File: CONTRIBUTING.md
Comment (kchibisov) [reply CONTRIBUTING.md:52]: No the `--` is a standard symbol for this. It's automatically added by the git itself when you author a patch for submission the _traditional_ way.

File: CONTRIBUTING.md
Comment (kchibisov) [reply CONTRIBUTING.md:36]: I know, but it's a problem of github in particular. Having correct links helps a lot when looking why the patch was send in the first place.

File: CONTRIBUTING.md
Comment (madsmtm) [reply CONTRIBUTING.md:52]: But you're writing that in the GitHub PR body, not in Git?

Comment (kchibisov) [reply CONTRIBUTING.md:52]: yeah, that's true. I'm just really used to that notation and it's quite common.

File: CONTRIBUTING.md
Comment (madsmtm) [reply CONTRIBUTING.md:159]: There's a whole set of different GitHub "roles", but some things do require admin permissions, take a look at https://github.com/rust-windowing/winit/settings/access.

Organization-wide, there's "Owner" and "Member", and only the latter can add others to the org.

In any case, I mostly wanted to codify our previous decision about discussing administrative changes with others before doing them.

**Review by daxpedda — CHANGES_REQUESTED** [review]
We decided to split off any commit message rules into a separate PR and merge everything else now, after all nits are resolved.

**Pull request closed without merging**
